Create cross-format `StandardProperties` class

NebelNidas opened the pull request:
Reduces redundancy and allows API consumers to have direct access to all standard properties.

NebelNidas committed:
Move Tiny properties to separate class

NebelNidas committed:
Remove duplicate `escaped-names` property

modmuss50 commented on src/main/java/net/fabricmc/mappingio/format/tiny/TinyProperties.java:
```diff
@@ -0,0 +1,36 @@
+ * All standard properties of the Tiny format.
+ * Internally, trees use the lowest common denominator.
+ */
+public final class TinyProperties {
```
Nit: Might want to add a private constructor to this since I merged: https://github.com/FabricMC/mapping-io/commit/806152ee3d81f7b1f4c0e8b2d3512d0f93e10530

Not that it really matters tho.

NebelNidas replied:
Will do in a few minutes! 👍

NebelNidas replied:
Fixed!

NebelNidas committed:
Prevent instantiation of `TinyProperties` class

NebelNidas commented:
Hmm, should I rename to class to `TinyFormatStandardProperties`?

NebelNidas committed:
Create cross-format `StandardProperties` class

- Now supports `component` intermediary counter
- Removes some hardcoded values
- Tiny v1 now supports reading arbitrary metadata (prefixed with `#`)

NebelNidas committed:
Fix checkstyle

NebelNidas committed:
Add missing license headers

NebelNidas committed:
Fix Tiny v1 metadata detection

NebelNidas committed:
Don't read properties from non-applicable formats

NebelNidas committed:
Use builder for standard properties

NebelNidas committed:
Consolidate builder functionality into `StandardPropertyImpl`

NebelNidas commented on src/main/java/net/fabricmc/mappingio/format/StandardProperty.java:
```diff
@@ -0,0 +1,34 @@
+
+import org.jetbrains.annotations.ApiStatus;
+
+public interface StandardProperty {
```
Probably yes, but since `MappingWriter` and the like aren't annotated as well, I'm not too sure

NebelNidas replied:
I've marked it as non-extendable for now, can always reconsider later.

NebelNidas commented on src/main/java/net/fabricmc/mappingio/format/StandardProperty.java:
```diff
@@ -0,0 +1,34 @@
+
+public interface StandardProperty {
+	Set<MappingFormat> getApplicableFormats();
+	boolean isApplicableTo(MappingFormat format);
```
Just a convenience method. Can remove if wanted

NebelNidas replied:
I've removed `getApplicableFormats()` now, since it's not strictly needed and allows for more future flexibility.

NebelNidas commented on src/main/java/net/fabricmc/mappingio/format/StandardProperties.java:
```diff
@@ -0,0 +1,138 @@
+	static {
+		NEXT_INTERMEDIARY_CLASS = builder("next-intermediary-class")
+				.add(MappingFormat.TINY_FILE, "INTERMEDIARY_COUNTER class")
+				.add(MappingFormat.TINY_2_FILE, "next-intermediary-class")
```
It's what mapping-io already uses internally: https://github.com/FabricMC/mapping-io/blob/master/src/main/java/net/fabricmc/mappingio/format/tiny/Tiny1FileReader.java#L183-L185

liach commented on src/main/java/net/fabricmc/mappingio/format/StandardProperties.java:
```diff
@@ -0,0 +1,138 @@
+	}
+
+	@ApiStatus.Internal
+	public static StandardProperty getById(String id) {
```
Say if you have a `next-intermediary-method` read from tiny v1, you would incorrectly interpret it as an intermediary counter while it isn't.

NebelNidas replied:
I've already added a sanity check preventing this: https://github.com/FabricMC/mapping-io/pull/29/commits/5341ecfb875845484f19c71a4c370842354c27be

NebelNidas replied:
The issue why a `Map<MappingFormat, Map<String, StandardProperty>>` doesn't work is due to intermediate operations, say `MemoryMappingTree.accept(MemoryMappingTree)`. I could generate a random ID per session, but IMO that's overcomplicating things without much reason

NebelNidas replied:
Hmm, not a huge fan of this, since up until this point the visitor- and tree-based APIs have always been cleanly separated

NebelNidas committed:
Merge upstream

NebelNidas committed:
Remove `getApplicableFormats` from `StandardProperty`

Don't expose more functionality than currently required, to reduce the API surface we need to maintain.

NebelNidas committed:
Mark `StandardProperty` as non-extendable

NebelNidas committed:
Allow to query for `StandardProperty`s via the tree API

NebelNidas committed:
Merge upstream

NebelNidas committed:
Add changes to changelog

NebelNidas committed:
Tiny v1 properly supports arbitrary metadata now

NebelNidas committed:
Merge upstream

NebelNidas committed:
Merge upstream

NebelNidas commented:
Oops, I wanted to request the review from Modmuss, sorry 😅

modmuss50 commented on src/main/java/net/fabricmc/mappingio/format/StandardProperties.java:
```diff
@@ -0,0 +1,111 @@
+		return valuesById.get(id);
+	}
+
+	public static final StandardProperty NEXT_INTERMEDIARY_CLASS;
+	public static final StandardProperty NEXT_INTERMEDIARY_FIELD;
+	public static final StandardProperty NEXT_INTERMEDIARY_METHOD;
+	public static final StandardProperty NEXT_INTERMEDIARY_COMPONENT;
```
Im not sure these should exist in mio, they are specific to intermediary and are generated by matcher. They arent part of the tiny format, and I dont think we actually read them anymore?

modmuss50 replied:
happy to discuss this if you think otherwise.

NebelNidas replied:
I'd say this is a bit of an edge-case, since the counters [have](https://github.com/FabricMC/mapping-io/blob/516a029c09d1354b15bf0a27e01f08e94c7c29c1/src/main/java/net/fabricmc/mappingio/format/tiny/Tiny1FileReader.java#L190-L192) [been](https://github.com/FabricMC/stitch/blob/0abaedb727630b679224667f3cc8950428db8f7d/src/main/java/net/fabricmc/stitch/commands/GenState.java#L489) special-cased until now and consequently behaved as de-facto standard properties. I could remove it, so we end up using the `INTERMEDIARY_COUNTER` notation in Tiny v2, too, though this would conflict with existing naming conventions and we would be stuck with a misleading name. We could of course also delegate this responsibility to Matcher, which would have to do the conversion via a custom forwarding visitor pass

NebelNidas replied:
In the long run we might need an API letting consumers register their own standard properties

NebelNidas committed:
Make Tiny v1 metadata detection more resilient against false-positives

liach commented on src/main/java/net/fabricmc/mappingio/format/StandardProperties.java:
```diff
@@ -0,0 +1,111 @@
+
+		private StandardPropertyImpl addMapping(MappingFormat format, String name) {
+			nameByFormat.put(format, name);
+			valuesByName.put(name, this);
```
Is there any potential scenario where a name refers to distinct properties in different formats? If yes, we should track (format, name) -> property mapping instead.

NebelNidas replied:
Not yet, but future-proofing for such a scenario doesn't hurt. Good catch!

liach commented on src/main/java/net/fabricmc/mappingio/format/StandardProperties.java:
```diff
@@ -0,0 +1,111 @@
+	}
+
+	public static Set<StandardProperty> values() {
+		return Collections.unmodifiableSet(values);
```
Is there a particular reason we should use a Set instead of just using `valuesByName.values()` or `valuesById.values()`?

NebelNidas replied:
Honestly, I don't remember. Might have been an oversight

liach reviewed (commented):
Also for preventing clash of property id with mapping's native properties, should we declare a new MappingFlag to alert mapping visitors? Also to preserve backward compatibility if say, an old tinyv1 visitor still anticipates old property keys instead of the standardized ones. If a visitor anticipates standardized keys, they should declare the new flag in their mapping flags.

Otherwise this whole thing looks pretty good, a huge improvement from the initial revision.

NebelNidas committed:
Address reviews

NebelNidas commented:
> to preserve backward compatibility if say, an old tinyv1 visitor still anticipates old property keys instead of the standardized ones. If a visitor anticipates standardized keys, they should declare the new flag in their mapping flags.

That's the nice thing, MIO [already used](https://github.com/FabricMC/mapping-io/blob/516a029c09d1354b15bf0a27e01f08e94c7c29c1/src/main/java/net/fabricmc/mappingio/format/tiny/Tiny1FileReader.java#L190-L192) the same standardized values internally before, so this whole PR is 100% backward compatible :)

NebelNidas commented:
Postponing to a later release, since I need to rewrite this to be more extensible
